File: src/Test/src/main/java/eu/modernmt/test/contextanalyzer/context/Stats.java
package eu.modernmt.test.contextanalyzer.context;

import eu.modernmt.context.ContextDocument;
import org.json.simple.JSONObject;

import java.util.List;

/**
 * Created by lucamastrostefano on 17/02/16.
 */
public class Stats {

    protected int numberOfMatches;
    protected int sumPosition;
    protected double sumScoreGap;
    protected int sumRecall;
    protected int numberOfQuery;
    protected long totalTime;

    public Stats(){
        this.numberOfMatches = 0;
        this.sumPosition = 0;
        this.sumScoreGap = 0;
        this.sumRecall = 0;
        this.numberOfQuery = 0;
        this.totalTime = 0;
    }

    public void addSample(List<ContextDocument> results, String domainId, long queryTime){
        this.numberOfQuery++;
        boolean match = results.get(0).getId().equals(domainId);
        boolean found = false;
        int position = -1;
        double gap = -1;
        for(int i = 0; i < results.size(); i++){
            if(results.get(i).getId().equals(domainId)){
                position = (i+1);
                if ((i+1) < results.size()){
                    gap = results.get(i).getScore() - results.get(i+1).getScore();
                } else{
                    gap = results.get(i).getScore();
                }
                found = true;
            }
        }
        this.add(match, position, gap, found, queryTime);
    }

    private synchronized void add(boolean match, int position, double gap, boolean found, long queryTime){
        if(match) {
            this.numberOfMatches++;
        }
        if(found) {
            this.sumRecall++;
            this.sumPosition += position;
            this.sumScoreGap += gap;
        }
        this.totalTime += queryTime;
    }

    @Override
    public String toString(){
        return "Precision: " + this.numberOfMatches /(double)this.numberOfQuery +
                "\tRecall: " + this.sumRecall/(double)this.numberOfQuery +
                "\tAvgPosition: " + this.sumPosition/(double)this.sumRecall +
                "\tAvgScoreGap: " + this.sumScoreGap/(double)this.sumRecall +
                "\tAvgQuertTime " + this.totalTime/(double)this.numberOfQuery;
    }

    public JSONObject getJson(){
        JSONObject jsonStats = new JSONObject();
        jsonStats.put("numberOfMatches", this.numberOfMatches /(double)this.numberOfQuery);
        jsonStats.put("Recall", this.sumRecall/(double)this.numberOfQuery);
        jsonStats.put("avgPosition", this.sumPosition/(double)this.sumRecall);
        jsonStats.put("avgScoreGap", this.sumScoreGap/(double)this.sumRecall);
        jsonStats.put("AvgQuertTime", this.totalTime/(double)this.numberOfQuery);
        return jsonStats;
    }
}

File: test/tests/context_analyzer_precision_and_recall_test/inc/Stats.py
import json


class Stats:

    def __init__(self):
        self.__number_of_matches = 0
        self.__sum_position = 0
        self.__sum_score_gap = 0
        self.__number_of_found = 0
        self.__number_of_queries = 0
        self.__total_time = 0

    def add_sample(self, results, domain_id, query_time):
        self.__number_of_queries += 1
        match = False
        position = -1
        gap = -1
        found = False
        if len(results) > 0:
            match = (results[0]['id'] == domain_id)
            for idx, result in enumerate(results):
                if result['id'] == domain_id:
                    position = idx + 1
                    if idx + 1 < len(results):
                        gap = result['score'] - results[idx + 1]['score']
                    else:
                        gap = result['score']
                    found = True
                    break
        self.add(match, position, gap, found, query_time)

    def add(self, match, position, gap, found, query_time):
        if match:
            self.__number_of_matches += 1
        if found:
            self.__number_of_found += 1
            self.__sum_position += 1
            self.__sum_score_gap += gap
        self.__total_time += query_time

    def get_stats(self):
        float_number_of_queries = float(self.__number_of_queries)
        float_number_of_found = float(self.__number_of_found)
        stats = {
            'Precision': float('nan') if float_number_of_queries == 0 else self.__number_of_matches / float_number_of_queries,
            'Recall':  float('nan') if float_number_of_queries == 0 else self.__number_of_found / float_number_of_queries,
            'Average position':  float('nan') if float_number_of_found == 0 else self.__sum_position / float_number_of_found,
            'Average score gap':  float('nan') if float_number_of_found == 0 else self.__sum_score_gap / float_number_of_found,
            'Average query time [s]':  float('nan') if float_number_of_queries == 0 else self.__total_time / float_number_of_queries
        }
        return  stats

    def __str__(self):
        return json.dumps(self.get_stats(), indent=4, separators=(',', ': '))